### PR TITLE
POC support over grpc

### PR DESCRIPTION
**Pull request by andymck**
Sibyl part of https://github.com/helium/miner/pull/1068

### CODE REVIEW

File: config/test.config [hunk 1]
```diff
@@ -11,7 +11,6 @@
         ]
     },
     {grpcbox, [
-%%            {client, #{channels => [{default_channel, [{http, "localhost", 10001, []}], #{}}]}},
             {servers,
```
Comment (jeffgrunewald) [inline config/test.config:14]: if the whole grpcbox config block is being removed from the main sys.config does the test.config file still need any of that config as well?

[the remark this answers is not in the document]
Comment (andymck) [reply config/test.config:14]: should be fine to remove, sibyl_ct_utils  sets the config per node

Comment (andymck) [reply config/test.config:14]: should be fine to remove, sibyl_ct_utils  sets the config per node

File: src/grpc/helium_stream_sc_follow_impl.erl [hunk 2]
```diff
@@ -28,71 +28,49 @@
...
-    %% tracks which SCs we have send a closed msg for
-    sc_closes_sent = [] :: list(),
-    %% tracks which SCs we have send a closing msg for
-    sc_closings_sent = [] :: list(),
-    %% tracks which SCs we have send a closable msg for
-    sc_closables_sent = [] :: list(),
-    %% tracks which SCs we have send a dispute msg for
-    sc_disputes_sent = [] :: list()
-}).
+-type handler_state() :: #{
```
Comment (jadeallenx) [inline src/grpc/helium_stream_sc_follow_impl.erl:31]: Why did you switch from a record to a map? Is this a situation where the internal keys will need to ever be serialized repeatably externally? (We will need to sort the keys in that case.)

File: src/grpc/helium_unary_general_impl.erl [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    Ctx,
+    #gateway_validators_req_v1_pb{
+        quantity = NumVals
+    } = Request
+) ->
+    lager:debug("executing RPC validators with msg ~p", [Request]),
+    %% get list of current validators from the cache
+    %% and then get a random sub set of these with size
+    %% equal to NumVals
+    IgnoreVals = application:get_env(sibyl, validator_ignore_list, []),
```
Comment (jadeallenx) [inline src/grpc/helium_unary_general_impl.erl:174]: We are going to end up needing a way to publish this information dynamically, right? We certainly don't want to necessarily do a software release for this

File: config/test.config (hunk 1 above)
[the remark this answers is not in the document]
Comment (andymck) [reply config/test.config:15]: actually the config is still required in test as it starts up a local node which requires this config

Comment (andymck) [reply config/test.config:15]: as per prev comment still required in test config

File: src/grpc/helium_unary_general_impl.erl (hunk 3 above)
Comment (andymck) [reply src/grpc/helium_unary_general_impl.erl:174]: Ignore vals here is not related to the deny list.  It is there to allow specified vals to be excluded from the list of validators returned to clients as part of the validators API.  This is useful if we know for example a validator which is running old code and we dont want it being used by clients.  Prob most useful on testnet where we have vals which just dont update.....no hard in leaving it in tho for time being

File: src/grpc/helium_stream_sc_follow_impl.erl (hunk 2 above)
Comment (andymck) [reply src/grpc/helium_stream_sc_follow_impl.erl:31]: switched to maps for convenience of accessing state in each of the different handlers.  The helium_gateway_service module needs to access the state in order to determine which modules to route info msgs to ( see https://github.com/helium/sibyl/blob/andymck/poc-grpc/src/grpc/helium_gateway_service.erl#L90 )

The use here is solely for internal process state

File: src/grpc/helium_gateway_service.erl [hunk 4]
```diff
@@ -9,179 +9,191 @@
...
-%%
-%% state channel unary APIs
-%%
-init(_RPC = is_active_sc, StreamState) ->
-    StreamState;
-init(_RPC = is_overpaid_sc, StreamState) ->
-    StreamState;
-init(_RPC = close_sc, StreamState) ->
+
+%% todo update to use stream_ prefix convention, requires gateway rs update
```
Comment (jeffgrunewald) [inline src/grpc/helium_gateway_service.erl:68]: does this have an issue or PR associated in gateway-rs yet to follow?

File: src/grpc/helium_stream_config_update_impl.erl [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    lager:debug("received config_update_notify msg, sending to client ~p", [Msg]),
+    %% received a config update notification event, we simply have to forward this unmodified to the client
+    %% the payload is fully formed and encoded
+    NewStreamState = grpcbox_stream:send(false, Msg, StreamState),
+    NewStreamState;
+handle_info(
+    _Msg,
+    StreamState
+) ->
+    lager:warning("unhandled info msg: ~p", [_Msg]),
```
Comment (jeffgrunewald) [inline src/grpc/helium_stream_config_update_impl.erl:56]:
```suggestion
    lager:debug("unhandled info msg: ~p", [_Msg]),
```

File: src/grpc/helium_stream_poc_impl.erl [hunk 6]
```diff
@@ -0,0 +1,163 @@
+-module(helium_stream_poc_impl).
+
+-include("../../include/sibyl.hrl").
```
Comment (jeffgrunewald) [inline src/grpc/helium_stream_poc_impl.erl:3]: do these header files in the `include/` dir need relative paths? i thought rebar knew to look for `.hrl` files in a root `include/` directory so these could just be `-include("sibyl.hrl").`

**Review by madninja — APPROVED** [review]

File: src/grpc/helium_stream_config_update_impl.erl [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    grpcbox_stream:t()
+) -> {ok, grpcbox_stream:t()} | grpcbox_stream:grpc_error_response().
+config_update(#gateway_config_update_req_v1_pb{} = Msg, StreamState) ->
+    Chain = sibyl_mgr:blockchain(),
+    #{streaming_initialized := StreamingInitialized} = grpcbox_stream:stream_handler_state(
+        StreamState
+    ),
+    config_update(Chain, StreamingInitialized, Msg, StreamState);
+config_update(_Msg, StreamState) ->
+    lager:warning("unhandled msg ~p", [_Msg]),
```
[the remark this answers is not in the document]
Comment (andymck) [reply src/grpc/helium_stream_config_update_impl.erl:40]: Added as I want to see how often and on what streams we get unexpected msgs.  Part of investigating the EOS issue as encountered on the follow sc stream

File: src/grpc/helium_stream_config_update_impl.erl [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    %% any process subscribed to it will receive chain var updates
+    Topic = sibyl_utils:make_config_update_topic(),
+    ok = sibyl_bus:sub(Topic, self()),
+    HandlerState = grpcbox_stream:stream_handler_state(StreamState),
+    NewStreamState = grpcbox_stream:stream_handler_state(
+        StreamState,
+        HandlerState#{
+            streaming_initialized => true
+        }
+    ),
```
[the remark this answers is not in the document]
Comment (andymck) [reply src/grpc/helium_stream_config_update_impl.erl:105]: In the above, we get the *existing* state and then update that.  On lines 23-26 as part of the init call there will be no existing state and thus we set only

File: src/grpc/helium_stream_poc_impl.erl (hunk 6 above)
Comment (andymck) [reply src/grpc/helium_stream_poc_impl.erl:3]: They should but yet it was failing to be found....may have been a rebar bug at the time

**Pull request merged**